File: 백준/Silver/6603. 로또/로또.java
import java.io.*;
import java.util.*;

public class Main {

    final static int MAX_COUNT = 6;
    static int[] s;
    static int k;
    static int[] d;
    static StringBuilder sb = new StringBuilder();

    public static void backtracking(int cnt, int last) {
        if (cnt == MAX_COUNT) {
            for (int i = 0; i < MAX_COUNT; i++) {
                sb.append(d[i]).append(' ');
            }
            sb.append('\n');
            return;
        }

        for (int i = last + 1; i < k; i++) {
            d[cnt] = s[i];
            backtracking(cnt + 1, i);
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        while (true) {
            st = new StringTokenizer(br.readLine());
            k = Integer.parseInt(st.nextToken());
            if (k == 0) break;
            s = new int[k];
            d = new int[6];

            for (int i = 0; i < k; i++) {
                s[i] = Integer.parseInt(st.nextToken());
            }
            backtracking(0, -1);
            sb.append('\n');
        }

        System.out.print(sb);
    }
}

File: 백준/Silver/6603. 로또/로또.kt
import java.util.*
/**
 * 1. 1~49중 k개의 숫자 고르기(주어짐)
 * 2. k개의 숫자 중 6개의 조합 나열하기
 * - 조합은 순서를 고려하면 안 됨
 * - 오름차순을 강제함으로써 순서 배제
 */

fun main() = with(System.`in`.bufferedReader()) {
    val sb = StringBuilder()
    while (true) {
        val st = StringTokenizer(readLine())
        val k = st.nextToken().toInt()
        if (k == 0) break

        val arr = IntArray(k) { st.nextToken().toInt() }
        val numbers = IntArray(6)

        fun comb(startArrIdx: Int, curNumberIdx: Int) {
            if (curNumberIdx == 6)  {
                sb.append(numbers.joinToString(separator = " ", postfix = "\n"))
                return
            }
            
            for (arrIdx in startArrIdx until k) {
                numbers[curNumberIdx] = arr[arrIdx]
                comb(arrIdx + 1, curNumberIdx + 1)
            }
        }

        comb(0, 0)
        sb.append('\n')
    }
    print(sb)
}